apps/api/src/app/events/usecases/add-job/add-digest-job.usecase.ts

```diff
@@ -1,55 +1,100 @@
-    if (!isValidDigestStep || !data) {
-      return undefined;
+  constructor(private jobRepository: JobRepository, protected createExecutionDetails: CreateExecutionDetails) {}
+
+  public async execute(command: AddDigestJobCommand): Promise<AddDigestJobResult> {
+    const { job } = command;
+
+    this.validateDigest(job);
+
+    return await this.shouldDelayDigestOrMerge(job);
+  }
+
+  private validateDigest(job: JobEntity): void {
+    if (job.type !== StepTypeEnum.DIGEST) {
+      throw new ApiException('Job is not a digest type');
     }
 
-    const shouldAddDigest = await this.shouldAddDigestJob.execute(command);
+    if (!job.digest?.amount) {
+      throw new ApiException('Invalid digest amount');
+    }
+
+    if (!job.digest?.unit) {
+      throw new ApiException('Invalid digest unit');
+    }
+  }
 
-    if (!shouldAddDigest) {
-      await this.createExecutionDetails.execute(
-        CreateExecutionDetailsCommand.create({
-          ...CreateExecutionDetailsCommand.getDetailsFromJob(data),
-          detail: DetailEnum.DIGEST_MERGED,
-          source: ExecutionDetailsSourceEnum.INTERNAL,
-          status: ExecutionDetailsStatusEnum.SUCCESS,
-          isTest: false,
-          isRetry: false,
-        })
-      );
+  private async shouldDelayDigestOrMerge(job: JobEntity): Promise<AddDigestJobResult> {
+    const { matched, modified } = await this.shouldDelayDigestOrMergeWithLock(job);
+
+    // We merged the digest job as there was an existing delayed digest job for this subscriber and template in the same time frame
+    if (matched > 0 && modified === 0) {
+      await this.digestMergedExecutionDetails(job);
 
       return undefined;
     }
 
-    await this.jobRepository.updateStatus(command.organizationId, data._id, JobStatusEnum.DELAYED);
-    if (!data.digest?.amount || !data.digest?.unit) {
-      throw new ApiException('Invalid digest amount or unit');
+    // We delayed the job and created the digest
+    if (matched === 0 && modified === 1) {
+      const { digest } = job;
+
+      if (!digest?.amount || !digest?.unit) {
+        throw new ApiException(`Somehow ${job._id} had wrong digest settings and escaped validation`);
+      }
+
+      return AddJob.toMilliseconds(digest.amount, digest.unit);
     }
 
-    return AddJob.toMilliseconds(data.digest.amount, data.digest.unit);
+    return undefined;
+  }
+
+  private async shouldDelayDigestOrMergeWithLock(job: JobEntity): Promise<IFindAndUpdateResponse> {
+    const resource = `environment:${job._environmentId}:template:${job._templateId}:subscriber:${job._subscriberId}`;
+    const TTL = 250;
+
+    const shouldDelayDigestDigestJobOrMerge = async () => this.jobRepository.shouldDelayDigestJobOrMerge(job);
+
+    return await applyLock<IFindAndUpdateResponse>(
+      {
+        resource,
+        ttl: TTL,
+      },
+      shouldDelayDigestDigestJobOrMerge
+    );
+  }
```

Comment:
Lock application. We surround the starting point of code we want to lock, so any call to that point of code, if the lock is taken, will be halted and retried once the lock is being freed.